.upstream-tests/test/std/atomics/atomics.flag/default.pass.cpp

```diff
@@ -25,26 +25,40 @@
 #endif
 #include "cuda_space_selector.h"
 
+
 template<template<typename, typename> class Selector>
 __host__ __device__
 void test()
 {
+// cudafe crashes on trying to interpret the line below when compiling with Clang
+// TODO: file a compiler bug
+#if !(defined(__clang__) && defined(__CUDACC__))
+# define _TEST_NO_DESTRUCT_ZERO
+#else
+# define _TEST_NO_DESTRUCT_ZERO zero.~A()
+#endif
+
     Selector<cuda::std::atomic_flag, default_initializer> sel;
     cuda::std::atomic_flag & f = *sel.construct();
     f.clear();
     assert(f.test_and_set() == 0);
     {
-#if !defined(__CUDA_ARCH__) || __CUDA_ARCH__ >= 700
-        typedef cuda::std::atomic_flag A;
-        TEST_ALIGNAS_TYPE(A) char storage[sizeof(A)] = {1};
-        A& zero = *new (storage) A();
-        assert(!zero.test_and_set());
-        // cudafe crashes on trying to interpret the line below when compiling with Clang
-        // TODO: file a compiler bug
-#if !(defined(__clang__) && defined(__CUDACC__))
-        zero.~A();
-#endif
-#endif
+        NV_DISPATCH_TARGET(
+            NV_PROVIDES_SM_70, (
+                typedef cuda::std::atomic_flag A;
+                TEST_ALIGNAS_TYPE(A) char storage[sizeof(A)] = {1};
+                A& zero = *new (storage) A();
+                assert(!zero.test_and_set());
+                _TEST_NO_DESTRUCT_ZERO;
+            ),
+            NV_IS_HOST, (
+                typedef cuda::std::atomic_flag A;
+                TEST_ALIGNAS_TYPE(A) char storage[sizeof(A)] = {1};
+                A& zero = *new (storage) A();
+                assert(!zero.test_and_set());
+                _TEST_NO_DESTRUCT_ZERO;
```

Comment:
Move into a lambda/named function and invoke that lambda/function on the branches to avoid code duplication?